packages/react-i18n/src/index.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * WordPress dependencies
+ */
+import React, {
+	createContext,
+	useContext,
+	useEffect,
+	useMemo,
+	useState,
+	useCallback,
+} from '@wordpress/element';
+import { createI18n, getLocaleData } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+
+const SCRIPT_HANDLES = [];
+
+const Context = createContext( null );
+
+export function LocaleProvider( { locale: initialLocale, domain, children } ) {
+	const [ locale, setLocale ] = useState( initialLocale );
+	const [ localeMap, setLocaleMap ] = useState( () => {
+		return new Map().set(
+			initialLocale,
+			createI18n( getLocaleData( domain ), domain )
+		);
+	} );
+
+	const switchToLocale = useCallback(
+		( newLocale ) => {
+			if ( ! localeMap.get( newLocale ) ) {
+				setLocaleMap(
+					new Map(
+						localeMap.set( newLocale, createI18n( {}, domain ) )
+					)
+				);
+			}
+			setLocale( newLocale );
+		},
+		[ localeMap ]
+	);
+
+	// Adds new locale data (= translations) and updates the map to force a re-render.
+	const setLocaleData = useCallback(
+		( data ) => {
+			localeMap.get( locale ).setLocaleData( data );
+			setLocaleMap(
+				new Map( localeMap.set( locale, localeMap.get( locale ) ) )
+			);
```

Comment:
This could be just `setLocaleMap( new Map( localeMap ) )`. The `set( get() )` bit doesn't do anything. We just need to update state with a new copy of the map that's not `===` to the previous one.